File: binarySearch/minInRotatedSortedArr.java
public int findMin(int[] nums){
    if(nums == null || nums.length == 0)
        return -1;
    int start = 0;
    int end = nums.length-1;
    int target = nums[nums.length-1];

    // find the first element(from left) that <= target
    // compare mid element and last element
    while(start + 1 < end){
        int mid = start + (end - start)/2;
        //right side is sorted, check left side
        if(nums[mid] <= target)
            end = mid;
        //left side is sorted, check right side
        else 
            start = mid;
    }
    //left+1=right:
    //arr[mid]=arr[left],
    //min=min(arr[mid],arr[right])
    if(nums[start] <= target)
        return nums[start];
    return nums[end];
}

File: binarySearch/rotated/minInRotatedSortedArr.py
def minRotatedSortedArr(arr):
    if arr is None or not arr:
        return -1

    left, right = 0, len(arr)-1

    while left + 1 < right:
        mid = left + (right - left)//2
        # duplicate
        if arr[mid] == arr[right]:
            right -= 1

        elif arr[mid] < arr[right]:
            right = mid

        elif arr[mid] > arr[right]:
            left = mid


    return min(arr[left], arr[right])


'''
 -
-
     -
    -
  - mid low

'''
-
    -
   -
  -
 -



    -
   -
  -
 -
      -
